### DISCUSSION TITLE
How bad am I at coding?

### DISCUSSION THREAD

**Discussion by ghost** [community/community#114308] [Programming Help]
```python
import os
line = 0
increment_line = True
show_line = True
code = []
cmds = ["view_array", "cleararray", "exit", "dis_inc_line", "enbl_inc_line", "dis_shw_line", "enbl_shw_line", "save_file", "fix", "goto", "clear", "set_line", "GUI_MODE", "gui_mode", "help"]
while True:
    if show_line: 
         cmd = input(str(line) + " > ")
    else:
         cmd = input("> ")
    if cmd == "exit":
        break 
    elif cmd.startswith("dis_inc_line"):
            increment_line = False
    elif cmd.startswith("enbl_inc_line"):
         increment_line = True
    elif cmd.startswith("dis_shw_line"):
         show_line = False
         increment_line = False
    elif cmd.startswith("enbl_shw_line"):
         show_line = True
         increment_line = True
    elif cmd == "save_file":
         file = input("Enter a file > ")
         if os.path.exists(file):
              with open(file, "a") as af:
                   for c in code:
                        af.append(code[c])
                        af.close()
                        print("Your code has been saved!")
         else:
              with open(file, "w") as wf:
                   for c in code:
                        wf.write(code[c])
                        wf.close()
                        print(f"The file {file} has been written!")
    elif cmd.startswith("fix"):
        index = int(cmd[3::])
        code.pop(index)
        instruction = input("Enter the new instruction > ")
        code.append(instruction)
    elif cmd.startswith("goto"):
         cmdAsInt = int(cmd[4::])
         code[cmdAsInt] = input("Enter the new instruction > ")
    elif cmd == "clear":
         os.system("cls")
    elif cmd.startswith("set_line"):
        line = int(cmd[8::])
        increment_line = False
    elif cmd == "help":
        print("The commands are:")
        for command in cmds:
            print(command)
        increment_line = False
    elif cmd == "GUI_MODE" or cmd == "gui_mode":
         print("Entering GUI Mode...")
         import GUIMode
         GUIMode.title()
    elif cmd == "viewarray":
        if len(code) > 0:
            for c in code:
                print(c)
        else:
            print("The array is empty!")
    elif cmd == "cleararray":
        code.clear()
    else:
         code.append(cmd)   
    if increment_line:
         line+=1

```
I also have some more code on my GitHub page: [My GitHub page](https://github.com/pyDOS8/)

**Comment by SiGMafEAr** [comment]
Practice makes perfect: Your code looks solid, but personally, I dig revising my code over and over to hack more efficient solutions. I mean, it's cool to get feedback from others, but nothing beats pushing our own limits for mad skillz and personal and professional leetsauce.

**Reply by ghost** [reply]
```cpp
#include <iostream>
#include <string>
using namespace std;
int main(){
	std::unique_ptr<int>interator = std::make_unique<int>(); // I hate positivity 
	std::unique_ptr<string>Operator = std::make_unqiue<string>(); // What a crappy variable name `Operator`
	std::unique_ptr<float(1000);>sums = new std::make_unqiue<float(1000)>();
	while(true){
		cout << "Enter an operator > ";
		cin >> *Operator;
		if(*Operator == "+"){
			std::unique_ptr<float>num1 = std::make_unique<float>();
			std::unique_ptr<float>num2 = std::make_unique<float>();
			cout << "Enter a number > ";
			cin >> *num1;
			cout << "Enter another number > ";
			cin >> *num2;
			cout << "The sum is " << num1+num2 << endl;
			for(int i = 0; i < 1000; i++){
				if(*sums[i] == 0.0){
					*sums[i] = *num1 + *num2;
				}else{
					*sums[i] = 0.0;
				}
			}
		}else if(*Operator == "-"){
			std::unique_ptr<float>num1 = std::make_unqiue<float>();
			std::unique_ptr<float>num2 = std::make_unique<float>();
			cout << "Enter a number > ";
			cin >> *num1;
			cout << "Enter another number > ";
			cin >> *num2;
			cout << "The result is " << *num1 - *num2 << endl;
			for(int i = 0; i < 1000; i++){
				if(*sums[i] == 0){
					*sums[i] = *num1 + *num2;
				}else{
					*sums[i] = 0.0;
				}
			}
		}else if(*Operator == "*"){
			std::unique_ptr<float>num1 = std::make_unqiue<float>();
			std::unique_ptr<float>num2 = std::make_unique<float>();
			cout << "Enter a number > ";
			cin >> *num1;
			cout << "Enter another number > ";
			cin >> *num2;
			cout << "The result is " << *num1 * *num2 << endl;
			for(*iterator = 0; *iterator < 1001; *iterator++){
				if(*sums[i] == 0.0){
					*sums[i] = *num1 * *num2;
				}else{
					*sums[i] = 0.0;
				}
			}
		}else if(*Operator == "/"){
			std::make_unqiue<float>num1 = std::make_unqiue<float>();
			std::make_unqiue<float>num2 = std::make_unique<float>();
			cout << "Enter a number > ";
			cin >> *num1;
			cout << "Enter another number > ";
			cin >> *num2;
			cout << "The result is " << *num1 / *num2 << endl;
			for(*iterator = 0; *iterator < 1001; *iterator++){
				if(*sums[i] == 0.0){
					*sums[i] = *num1 / *num2;
				}else{
					*sums[i] = 0.0;
				}
			}
		}else{
			cout << *Operator << " is not an existing operator!" << endl;
		}
	}
	return 0;
}
``` 
What about my C++ skills? I think they're lacking due to my lack of intelligence.

**Reply by ghost** [reply]
I used smart pointers to reduce memory and have safe memory interactions because they automatically deallocate the memory for me.

**Reply by SiGMafEAr** [reply]
Lots of folks struggle with pointers in C++... I totally borked their implementation at first, but once I grokked them, they boosted memory management and code execution big time.

**Reply by SiGMafEAr** [reply]
#include <iostream>
#include <string>
#include <vector>
using namespace std;

int main() {
    vector<float> sums(1000, 0.0); // Vector to store sums
    char Operator; // Operator input variable
    
    while (true) {
        cout << "Enter an operator (+, -, *, /) > ";
        cin >> Operator;

        if (Operator == '+' || Operator == '-' || Operator == '*' || Operator == '/') {
            float num1, num2;
            cout << "Enter a number > ";
            cin >> num1;
            cout << "Enter another number > ";
            cin >> num2;

            float result;
            switch (Operator) {
                case '+':
                    result = num1 + num2;
                    break;
                case '-':
                    result = num1 - num2;
                    break;
                case '*':
                    result = num1 * num2;
                    break;
                case '/':
                    if (num2 != 0) {
                        result = num1 / num2;
                    } else {
                        cout << "Error: Division by zero!" << endl;
                        continue; // Restart the loop
                    }
                    break;
            }

            cout << "The result is " << result << endl;

            // Store the result in the sums vector
            for (int i = 0; i < 1000; ++i) {
                if (sums[i] == 0.0) {
                    sums[i] = result;
                    break; // Exit loop after storing the result
                }
            }
        } else {
            cout << Operator << " is not a valid operator!" << endl;
        }
    }

    return 0;
}

/*++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Personally, I would've sorted it using the following code.

Check it out, here's the scoop:
We ditched those unique pointers and went for a basic vector to stash the sums, keeping it simple and sidestepping dynamic memory.
Switched the operator's data type to char, 'cause who needs more than one character to handle an operator?
Threw in a switch loop to crunch the operator input and bust out the right moves.
Slapped on a safety check to steer clear of zero division disasters.
Cut out all the redundant unique pointer jazz and memory fuss.
Tweaked those error and input messages for smoother sailing.
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*/
